### ISSUE TITLE
Test upcoming Twine release (1.10.0rc1)?

### ISSUE THREAD

**Issue by brainwane** [devpi/devpi#520]
https://test.pypi.org/project/twine/1.10.0rc1/  is a release candidate for Twine 1.10.0 which I'm planning to release early this week. Please see the [changelog](https://twine.readthedocs.io/en/latest/changelog.html) for detailed notes under "Next feature release".

I believe this is how you test it out on its own:  `pip install --upgrade --pre --index-url https://test.pypi.org/simple/ --extra-index-url https://pypi.org/simple twine` 

I would to hear about any problems this new release causes for `devpi`. And if there are some simple tests I can run before a new release to check whether a change in `twine` would break core functionality for you, maybe you could mention that on pypa/twine#314.

Thanks.

**Comment by hpk42** [comment]
Haven't used twine in a while (we have "devpi upload" which does a lot of what twine does) but now tried the 1.10.0rc1 -- i could use it fine with uploading to pypi.org and then tried with an index on devpi.  I had to install the "keyrings.alt" package to be able to use basic auth but it failed later then: 

  hpk@beto:~/p/muacrypt (master)$ twine upload --repository-url https://m.devpi.net/hpk/dev dist/muacrypt-0.8.0.dev6*
  Uploading distributions to https://m.devpi.net/hpk/dev
  Enter your username: hpk
  Enter your password: 
  Uploading muacrypt-0.8.0.dev6-py2.py3-none-any.whl
  100%|█████████████████████████████████████████████████| 37.9k/37.9k [00:00<00:00, 70.7kB/s]
  HTTPError: 400 Client Error: Bad Request for url: https://m.devpi.net/hpk/dev

As i am a little ill, i don't pursue this further right now -- maybe @fschulze or @gotcha can look into it but i guess not today.

**Comment by jaraco** [comment]
Seems to be working for me.

```
$ ~/.envs/twine-pr/bin/twine upload -r devpi *
Uploading distributions to https://devpi.net/jaraco/dev/
Uploading pkg_resources-0.3.0-py2.py3-none-any.whl
100%|█████████████████████████████████████████████████████████████████████████████████████| 137k/137k [00:01<00:00, 12.2kB/s]
Uploading pkg_resources-0.3.0.tar.gz
100%|██████████████████████████████████████████████████████████████████████████████████████| 128k/128k [00:00<00:00, 159kB/s]
```

**Comment by jaraco** [comment]
> I had to install the "keyrings.alt" package to be able to use basic auth.

Ugh. The keyring support should fail gracefully if there's no keyring available. I can help address that.

**Comment by jaraco** [comment]
I've added pypa/twine#315 to address the issue.

**Comment by brainwane** [comment]
Thanks for the advice and the testing. https://pypi.org/project/twine/1.10.0/ is now out. I don't have enough information to track down the issue @hpk42 was having, and it sounds like it might be local to his setup, so I'm going to close this issue, but please do file issues in the Twine repo if it turns out there's something broken on our side, or you suspect there is!

And I'd appreciate a note on pypa/twine#314 to help me understand what automated (or manual) tests a Twine maintainer could perform next time to check that the next release works with devpi.

Thanks again.

**Comment by mhuin** [comment]
@hpk42 I get the same 400 error trying to upload a wheel with twine on a devpi server. The logs don't say much more than "could not decode JSON".

2018-12-18 09:50:37,319 INFO  [req8] POST /testpi/dev
2018-12-18 09:50:37,319 DEBUG [req8] [Wtx4] in-transaction 4
2018-12-18 09:50:37,320 ERROR [req8] [Wtx4] while handling http://localhost:4040/testpi/dev:
Bad request: could not decode json

Have you found a fix for this?

**Comment by fschulze** [comment]
I just ran into a related issue. Please try with a trailing slash on the submit URL, like this: ``http://localhost:4040/testpi/dev/``
I will look into creating a proper fix.

**Comment by mhuin** [comment]
@fschulze yes, adding the slash solved the issue, thanks!

**Comment by fschulze** [comment]
A fix for this was released with devpi-server 4.8.1.
